### PR TITLE
mgmt: mcumgr: grp: img_mgmt: Add optional max image size reduction

### LINKED ISSUE #46194: mgmt/mcumgr/lib: Image upload should reject image that is too big to mcuboot to swap
**Is your enhancement proposal related to a problem? Please describe.**
Mcuboot may not be able to swap some of images, depending on compiled in upgrade mode, when image size does not leave some additional pages in primary image slot.
Mcumgr is not aware of mcuboot mode so it will just upload any image, to secondary slot, that will fit into primary slot, event if taking all free pages of that slot.
On restart mcuboot would not be able to swap image that is too big but user will not know this until upload is processed and reset is triggered, to perform upgrade.

**Describe the solution you'd like**
mcumgr update should be able reject images that would not be swappable  by mcuboot.
Kconfig options should be added to allow user to compile mcumgr for the same mode of operation the mcuboot is compiled for and logic should be added to check if image, an user tries to upload, will fit the mode.

**Describe alternatives you've considered**
1) Leaving as is: this issue can found during tests, prior to deployment, which means that such check would only take flash size,  while should be detected "in factory";
2) Using different slot sizes: making secondary slot shorter than primary, by required number of pages, will prevent mcumgr from uploading image that is too big to be swapped; problem is that different modes of operations of mcuboot would required different DTS overlays to be selected;
3) Kconfig options that would compile mcumgr to the same mode of operation as mcuboot it is going to work with;
then the mcumgr can decide whether it wants to accept the image or not, or which DTS partition scheme would be used.

**Additional context**
TODO

### PR DESCRIPTION
**Pull request by nordicjm**
Adds an optional feature that can be used to reduce the maximum allowed image upload file size whereby an image could be uploaded that would be too large to swap even if it could fit the partition

Fixes #46194

### CODE REVIEW

File: subsys/mgmt/mcumgr/grp/img_mgmt/Kconfig [hunk 1]
```diff
@@ -153,6 +153,38 @@ config MCUMGR_GRP_IMG_MUTEX
...
+	default MCUMGR_GRP_IMG_TOO_LARGE_DISABLED
+	help
+	  MCUboot images should be limited to the maximum size that the bootloader can swap, in
+	  order to know this size, additional information is needed from the MCUboot
+	  configuration, otherwise an image can be uploaded that is too large for the bootloader
+	  to swap, this selects which method to use.
+
+	  Note: setting this to a non-disabled option will prevent uploading of padded and
+	  confirmed images, if support for that is required then this feature should be left as
+	  disabled.
```
Comment (de-nordic) [inline subsys/mgmt/mcumgr/grp/img_mgmt/Kconfig:167]: Would it be possible for us to develop MCUboot/bootutil function that would check the image for valid footer?

File: subsys/mgmt/mcumgr/grp/img_mgmt/Kconfig [hunk 2]
```diff
@@ -153,6 +153,38 @@ config MCUMGR_GRP_IMG_MUTEX
...
+	bool "Via Sysbuild from MCUboot configuration"
+	depends on ROM_END_OFFSET > 0
+	help
+	  Will use the image overhead size calculated during Sysbuild image configuration.
+
+config MCUMGR_GRP_IMG_TOO_LARGE_BOOTLOADER_INFO
+	bool "Via retention bootloader info"
+	depends on RETENTION_BOOTLOADER_INFO_OUTPUT_FUNCTION
+	help
+	  Will fetch the maximum image size from the bootloader info retention subsystem module.
```
Comment (de-nordic) [inline subsys/mgmt/mcumgr/grp/img_mgmt/Kconfig:184]: We should be able to, also, provide this info at compile time. At all we are able to calculate space needed for footer in MCUboot, when we know size of slot and write block size.

**Review by de-nordic — DISMISSED** [review]

File: subsys/mgmt/mcumgr/grp/img_mgmt/Kconfig (hunk 2 above)
Comment (nordicjm) [reply subsys/mgmt/mcumgr/grp/img_mgmt/Kconfig:184]: This is provided at compile time when using the sysbuild option. If not using sysbuild then you cannot get that information

File: subsys/mgmt/mcumgr/grp/img_mgmt/Kconfig (hunk 1 above)
Comment (nordicjm) [reply subsys/mgmt/mcumgr/grp/img_mgmt/Kconfig:167]: Not possible, well, depending on what you mean by valid footer. Only MCUboot can confirm the image properly due to possibly being encrypted, an application could do some minor validation but nothing more than that

Comment (de-nordic) [reply subsys/mgmt/mcumgr/grp/img_mgmt/Kconfig:167]: Actually you can use imgtool to pad, confirm and encrypt image. 
But yes, I guess you can not check the footer if the MCUboot only knows the key.

Comment (nordicjm) [reply subsys/mgmt/mcumgr/grp/img_mgmt/Kconfig:167]: Also can't check that here because to check that you need to receive the whole image

Comment (de-nordic) [reply subsys/mgmt/mcumgr/grp/img_mgmt/Kconfig:167]: The image slot trailer, the one that stores swap bits, etc, is not encrypted. actually.

**Review by de-nordic — APPROVED** [review]

File: subsys/mgmt/mcumgr/grp/img_mgmt/Kconfig (hunk 1 above)
Comment (nordicjm) [reply subsys/mgmt/mcumgr/grp/img_mgmt/Kconfig:167]: Not encrypted no but only mcuboot can verify that it is a valid image or supported on that device, the application can't know anything more than the basic validity of parts of the image

**Review by fabiobaltieri — APPROVED** [review]

**Pull request merged**